app/code/Magento/Inventory/Model/Reservation.php

```diff
@@ -0,0 +1,66 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\Inventory\Model;
+
+use Magento\Framework\Model\AbstractModel;
+use Magento\Inventory\Model\ResourceModel\Reservation as ReservationResourceModel;
+use Magento\InventoryApi\Api\Data\ReservationInterface;
+
+/**
+ * {@inheritdoc}
+ *
+ * @codeCoverageIgnore
+ */
+class Reservation extends AbstractModel implements ReservationInterface
+{
+    /**
+     * @inheritdoc
+     */
+    protected function _construct()
+    {
+        $this->_init(ReservationResourceModel::class);
+    }
```

Comment:
will create an internal ticket to get this method marked as deprecated before the 2.2.0 Magento release

Reply:
Does this mean that (at least at the moment) we can get rid of `Magento\Inventory\Model\ResourceModel\Reservation` class?

Reply:
If I get the point, I can turn `Reservation` into a framework-agnostic domain model with its own `__constructor()`, without inheriting from `AbstractModel`.

At this point calling the `create()` method on the object manager gives an issue related on how the arguments array is passed.

In the `ReservationBuilder` I create a `Reservation` instance this way:

        $data = [
            ReservationInterface::RESERVATION_ID => $this->reservationId,
            ReservationInterface::STOCK_ID => $this->stockId,
            ReservationInterface::SKU => $this->sku,
            ReservationInterface::QUANTITY => $this->quantity,
            ReservationInterface::METADATA => $this->metadata,
        ];

        $arguments = [
            'data' => $data,
        ];

        return $this->objectManager->create(ReservationInterface::class, $arguments);

For some reasons, if I don't inherit from `AbstractModel` this call will fail because arguments cannot be resolved anymore in `Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgumentsInRuntime()`. 

It only works if I change the code as shown below:

        $arguments = [
            'reservationId' => $this->reservationId,
            'stockId' => $this->stockId,
            'sku' => $this->sku,
            'quantity' => $this->quantity,
            'metadata' => $this->metadata,
        ];

        return $this->objectManager->create(ReservationInterface::class, $arguments);

Maybe we can examine how to overcome this together?

Reply:
>Does this mean that (at least at the moment) we can get rid of Magento\Inventory\Model\ResourceModel\Reservation class?

It dependes whether we want to use Reservation Collection or not.
At least for now we don't have such use-cases

Reply:
>If I get the point, I can turn Reservation into a framework-agnostic domain model with its own __constructor(), without inheriting from AbstractModel.

Yes, exactly. The intention is totally correct.

Reply:
>For some reasons, if I don't inherit from AbstractModel this call will fail because arguments cannot be resolved anymore in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgumentsInRuntime().

I see your problem. The reason of the issue is how ObjectManager implemented in Magento 2.
When you pass Entity and $arguments for this entity to ObjectManager `create()` function, like this:
`$this->objectManager->create(ReservationInterface::class, $arguments);`
ObjectManager factory does next (simplified version):
```php
    public function create($requestedType, array $arguments = [])
    {
        // Get All the possible Arguments for Requested Type using Reflection 
        $args = $this->config->getArguments($requestedType);
        // Make binding between Interface -> Class to implement for requested type
        $type = $this->config->getInstanceType($requestedType);
            //looping through the all possible arguments for Requested type and 
            //if the argument with the same Name provided for ObjectManager::create - 
            //pass one for object being instantiated
            foreach ($args as $key => &$argument) {
                if (isset($arguments[$key])) {
                    $argument = $arguments[$key];
                } 
           }
```

What's important here, that argument resolver bind Entity's constructor parameters using  the same names (keys of array) in the $arguments array (provided as the second parameter for ObjectManager::create function).

In your case, inheriting from the `Magento\Framework\Model\AbstractModel`
Your Reservation class inherits parent's constructor as well. 
In this case, it looks like:
```php
    public function __construct(
        \Magento\Framework\Model\Context $context,
        \Magento\Framework\Registry $registry,
        \Magento\Framework\Model\ResourceModel\AbstractResource $resource = null,
        \Magento\Framework\Data\Collection\AbstractDb $resourceCollection = null,
        array $data = []
    ) {
```
The most interesting in our case is `$data array`.
Thus making a call:
```php
    $arguments = [
        'data' => $data,
    ];

    return $this->objectManager->create(ReservationInterface::class, $arguments);
```
ObjectManager makes a mapping between ` 'data' => $data,` and `array $data`
Thus at the time of Reservation instantiation ObjectManager will provide the whole array of data you prepared for the last constructor parameter ($data).
After that in the basic constructor these data would be assigned to data array
```php
    public function __construct(array $data = [])
    {
        $this->_data = $data;
    }
```

Removing inheritance from the AbstractModel - you lose the basic constructor.
Thus there is no $data parameter anymore in Reservation constructor, so data you provide can't be binded.

What you need to do - is to introduce own constructor for reservation accepting each of the argument as an independent parameter. Like this:
```php
class Reservation
{
    public function __construct(
         $stockId,
         $sku,
         $qty,
         $metadata = null
    ) {
}
```
after that you can make initialization this way:
```php
    $arguments = [
        ReservationInterface::STOCK_ID => $this->stockId,
        ReservationInterface::SKU => $this->sku,
        ReservationInterface::QUANTITY => $this->quantity,
        ReservationInterface::METADATA => $this->metadata,
    ];

    return $this->objectManager->create(ReservationInterface::class, $arguments);
```
each of the parameters would be bind independently in this case (not in the scope of $data array).